docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md

```diff
@@ -0,0 +1,217 @@
+gep-number: 15
+creation-date: 2021-03-31
+status: implementable
+authors:
+- "@petersutter"
+reviewers:
+- "@rfranzke"
+---
+
+# GEP-15: Bastion Management and SSH Key Pair Rotation
+
+## Table of Contents
+
+- [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+    - [Involved Components](#involved-components)
+    - [SSH Flow](#ssh-flow)
+    - [Resource Example](#resource-example)
+- [SSH Key Pair Rotation](#ssh-key-pair-rotation)
+    - [Rotation Proposal](#rotation-proposal)
+
+## Motivation
+`gardenctl` (v1) has the functionality to setup SSH sessions to the targeted shoot cluster. For this, infrastructure resources like vms, firewall rules etc. have to be created. `gardenctl` will clean up the resources after the SSH session. However, there were issues in the past where that infrastructure resources did not get cleaned up properly, for example due to some error and was not retried. Hence, the proposal is to have a dedicated controller (for each infrastructure) that manages the infrastructure resources. `gardenctl` also re-used the SSH node credentials for the bastion host. Instead, a new temporary SSH key pair should be created for the bastion host.
+The static shoot-specific SSH key pair should be rotated regularly, for example once in the maintenance time window.
+
+Changes to the `Bastion` resource should only be allowed for controllers on seeds that are responsible for it. This cannot be restricted when using custom resources.
+The proposal, as outlined below, suggests to implement the necessary changes in the gardener core components and to adapt the [SeedAuthorizer](https://github.com/gardener/gardener/issues/1723) to consider `Bastion` resources that the Gardener API Server serves.
+
+
+### Goals
+- Operators can request and will be granted **one-time**, **personal** , and **time-limited** `ssh` access to cluster nodes:
+  - **one-time**: Credentials are under Gardener control and can be used only for the extent of this request (of course multiple `ssh` sessions are possible, in parallel or repeatedly, but after "the time is up", see time-limited property, they will be revoked and can never be used ever again; this, btw., prohibits use of private credentials that Gardener cannot permanently revoke/rotate)
+  - **personal**: Credentials are bound to the requestor and logins and logs can be properly audited and traced back to the requestor
+  - **time-limited** Credentials are valid only for a specific time span and cannot be used beyond that time span
+- The respective generated **personal `ssh` credentials will be made available to the requestor** and nobody else can use (a.k.a. impersonate) them (not even other operators).
+- Necessary infrastructure resources for `ssh` access (such as VMs, firewall rules, etc.) are automatically created and also terminated after usage, but at the latest after the above mentioned max time span.
+
```

Comment:
```suggestion
- Operators can request and will be granted time-limited `ssh` access to shoot cluster nodes via bastion hosts.
- To that end, requestors must present their public `ssh` key and only this will be installed into `sshd` on the bastion hosts.
- The bastion hosts will be firewalled and ingress traffic will be permitted only from the client IP of the requestor.
- The actual node `ssh` private key (resp. key pair) will be rotated by Gardener and access to the nodes is only possible with this constantly rotated key pair and not with the personal one that is used only for the bastion host.
- Bastion host and access is granted only for the extent of this operator request (of course multiple `ssh` sessions are possible, in parallel or repeatedly, but after "the time is up", access is no longer possible).
- By these means (personal public key and allow-listed client IP) nobody else can use (a.k.a. impersonate) the requestor (not even other operators).
- Necessary infrastructure resources for `ssh` access (such as VMs, public IPs, firewall rules, etc.) are automatically created and also terminated after usage, but at the latest after the above mentioned time span is up.
```